prov/efa: use shm provider to support cuda buffer

wzamazon opened the pull request:
This PR contain 3 patches that fix several issues when using shm provider to support cuda buffer:

1.  request the FI_HMEM capability in cuda build
2. fix a bug the wrong descriptor was used for shm provider
3. do not use eager protocol for cuda buffer and local peer.

Signed-off-by: Wei Zhang <wzam@amazon.com>

wzamazon commented:
bot:aws:retest

zhngaj commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -179,6 +179,46 @@ ssize_t rxr_read_mr_reg(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
+ * @param efa_desc[out] descriptors for lower efa provider.
+ * @param shm_desc[out] descriptors for lower shm provider.
+ * @param desc[in] descriptors provided by application
+ * @param numdesc[in] number of descriptors in the array
```
Is @param low_ep_type missing here?
Also I think @desc_out should be here, and you can put the description of efa_desc and shm_desc under it.

wzamazon replied:
thank you! fixed in new revision

rwespetal commented on prov/efa/src/rxr/rxr_read.c:
```diff
@@ -179,6 +179,46 @@ ssize_t rxr_read_mr_reg(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
 }
 
+/**
+ * @brief convert descriptor from application for lower proivder to use
```
There's a typo here and on line 185

wzamazon replied:
Thanks! Fixed in new revision.

rwespetal reviewed (commented):
Did you scan through the other send paths where application descriptors could be used? I think I see some paths besides read where we'll try to use `efa_mr` with SHM. I also see that in `rxr_pkt_entry_send_with_flags` we'll avoid the descriptor entirely when using SHM which won't work if we're sending from a buffer registered with FI_HMEM.

wzamazon commented:
@rwespetal 

Yes, the original patch only fixed read and did not fix send path. 

I just updated the patch so send is included. Even with that patch, we still have problem with shm because it does not support mixed memory type of iov. I opened an issue for it (https://github.com/ofiwg/libfabric/issues/6639). 
To mitigate the issue, I added a patch that will not use eager protocol for cuda buffer on same host.
Finally, I added a patch that request FI_HMEM capability from shm provider. 
So, altogether there are 3 patches in this PR now. Please take a look!

rwespetal commented on prov/efa/src/rxr/rxr_pkt_entry.c:
```diff
@@ -362,13 +362,18 @@ ssize_t rxr_pkt_entry_sendv(struct rxr_ep *ep,
 	msg.context = pkt_entry;
 	msg.data = 0;
 
+	if (peer->is_local) {
```
Perhaps this should be in `rxr_pkt_entry_sendmsg`, since that's where fi_sendmsg is called? Looks like nothing calls sendmsg directly at the moment but that could change.

rwespetal replied:
Is it possible to make this change? Otherwise it looks fine to me.

wzamazon replied:
Move it to `rxr_pkt_entry_sendmsg` would require changing the function's signature by dropping the `const` prefix to msg, because we will need to modify `msg->desc` inside the function. I am open to it, but wonder if you are OK with that?

rwespetal replied:
Oh, I see. In that case, I'd just add a comment to `rxr_pkt_entry_sendmsg` that the caller must set `msg->desc` correctly based on whether this going over SHM, and to see sendv for an example.

rwespetal reviewed (commented):
Looks like the CI failures are real. The patches look fine to me so I must be missing something.

wzamazon commented:
@wesper The CI failure is indeed real. The issue was that EFA also uses shm provider's write and atomic, and is passing the wrong descriptors too. This has not caused issue before because EFA did not request FI_HMEM capability, so shm does not use descriptor.

Now that EFA requested FI_HMEM from shm, shm need to use the descriptors to identify memory type, hence the error.

I split the bug fixes into 4 patches: one for send, one for read, one for write and one for atomic. The fabtests passes now. 

The current AWS CI failure is because a transient GPU hardware issue.

wzamazon commented:
Sure, will add that in the next revision.

wzamazon commented:
@rwespetal I added a comment in new revision. I think it is better to merge `rxr_pkt_entry_sendv` and `rxr_pkt_entry_sendmsg` into one function, but I think that would be a refactor and need to be a separate commit.

rwespetal commented:
The comment is fine for now, I just wanted to make sure that point doesn't get missed.
